snapper/BtrfsUtils.cc

```diff
@@ -67,30 +70,79 @@ namespace snapper
 	bool
 	is_subvolume_read_only(int fd)
 	{
+#ifdef HAVE_LIBBTRFSUTIL
+	    enum btrfs_util_error err;
+	    bool readonly;
+
+	    err = btrfs_util_get_subvolume_read_only_fd(fd, &readonly);
+	    if (err)
+		throw runtime_error_with_errno("btrfs_util_get_subvolume_read_only_fd() failed", errno);
+
+	    return readonly;
+#else
 	    __u64 flags;
 	    if (ioctl(fd, BTRFS_IOC_SUBVOL_GETFLAGS, &flags) < 0)
 		throw runtime_error_with_errno("ioctl(BTRFS_IOC_SUBVOL_GETFLAGS) failed", errno);
 
 	    return flags & BTRFS_SUBVOL_RDONLY;
+#endif
 	}
 
 
 	void
 	create_subvolume(int fddst, const string& name)
 	{
+#ifdef HAVE_LIBBTRFSUTIL
+	    enum btrfs_util_error err;
+
+	    err = btrfs_util_create_subvolume_fd(fddst, name.c_str(), 0, NULL, NULL);
+	    if (err)
+		throw runtime_error_with_errno("btrfs_util_create_subvolume_fd() failed", errno);
+#else
 	    struct btrfs_ioctl_vol_args args;
 	    memset(&args, 0, sizeof(args));
 
 	    strncpy(args.name, name.c_str(), sizeof(args.name) - 1);
 
 	    if (ioctl(fddst, BTRFS_IOC_SUBVOL_CREATE, &args) < 0)
 		throw runtime_error_with_errno("ioctl(BTRFS_IOC_SUBVOL_CREATE) failed", errno);
+#endif
 	}
 
 
 	void
 	create_snapshot(int fd, int fddst, const string& name, bool read_only, qgroup_t qgroup)
 	{
+#ifdef HAVE_LIBBTRFSUTIL
+	    enum btrfs_util_error err;
+	    struct btrfs_util_qgroup_inherit *util_inherit = NULL;
+	    int flags = 0;
+
+	    if (read_only)
+		flags |= BTRFS_UTIL_CREATE_SNAPSHOT_READ_ONLY;
+
+#ifdef ENABLE_BTRFS_QUOTA
+	    size_t size = sizeof(btrfs_qgroup_inherit) + sizeof(((btrfs_qgroup_inherit*) 0)->qgroups[0]);
+	    vector<char> buffer(size, 0);
+
+	    if (qgroup != no_qgroup)
+	    {
+		struct btrfs_qgroup_inherit *inherit;
+
+		inherit = (btrfs_qgroup_inherit*) &buffer[0];
+		inherit->num_qgroups = 1;
+		inherit->num_ref_copies = 0;
+		inherit->num_excl_copies = 0;
+		inherit->qgroups[0] = qgroup;
+		util_inherit = (struct btrfs_util_qgroup_inherit *)inherit;
+	    }
+#endif
+	    err = btrfs_util_create_snapshot_fd2(fd, fddst, name.c_str(), flags, NULL, util_inherit);
+	    if (err && errno != ENOTTY && errno != EINVAL)
+		throw runtime_error_with_errno("btrfs_util_create_snapshot_fd2() failed", errno);
+
+	    /* No BTRFS_IOC_SNAP_CREATE fallback */
```

Comment:
AFAIS the fallback does not use ```BTRFS_IOC_SNAP_CREATE``` since the second ```#endif``` below is too late.

Whether the fallback is needed is another question (libbtrfsutil seems to be a few years younger than ```BTRFS_IOC_SNAP_CREATE_V2```).